readthedocs/doc_builder/environments.py

```diff
@@ -267,7 +270,86 @@ def get_wrapped_command(self):
                                    for part in self.command]))))
 
 
-class BuildEnvironment(object):
+class BaseEnvironment(object):
+
+    """
+    Base environment class.
+
+    Used to run arbitrary commands outside a build.
+    """
+
+    def __init__(self, project, environment=None):
+        # TODO: maybe we can remove this Project dependency also
+        self.project = project
+        self.environment = environment or {}
+        self.commands = []
+
+    def record_command(self, command):
+        pass
+
+    def _log_warn_only(self, msg):
+        log.warn(LOG_TEMPLATE.format(
+            project=self.project.slug,
+            version='latest',
+            msg=msg,
+        ))
+
+    def run(self, *cmd, **kwargs):
+        """Shortcut to run command from environment."""
+        return self.run_command_class(cls=self.command_class, cmd=cmd, **kwargs)
+
+    def run_command_class(self, cls, cmd, record=None, warn_only=False, **kwargs):
+        """
+        Run command from this environment.
+
+        :param cls: command class to instantiate a command
+        :param cmd: command (as a list) to execute in this environment
+        :param warn_only: don't raise an exception on command failure
+        :param record: whether or not to record this particular command
+        """
+        if record is None:
+            # ``self.record`` only exists when called from ``*BuildEnvironment``
+            record = getattr(self, 'record', False)
```

Comment:
If we do that, once you initialize the class with `False` there is no way to change the behaviour for a specific command. I supposed that we _want_ to override the default value for a specific command.

This would be confusing from my point of view:

1. initialize the class with `False`
2. call this method with `record=True`
3. it won't be recorded :/

To me, it's the same (opposite) case as "initialize the class with `True` but _this_ and _this_ command I don't want to record, so I call it `False`"